pkg/controller/daemon/daemoncontroller.go

```diff
@@ -1016,6 +1016,30 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 	return dsc.updateDaemonSetStatus(ds)
 }
 
+// hasIntentionalPredicatesReasons checks if any of the given predicate failure reasons
+// is intentional.
+func hasIntentionalPredicatesReasons(reasons []algorithm.PredicateFailureReason) bool {
+	for _, r := range reasons {
+		switch reason := r.(type) {
+		case *predicates.PredicateFailureError:
+			switch reason {
+			// intentional
+			case
+				predicates.ErrNodeSelectorNotMatch,
+				predicates.ErrPodNotMatchHostName,
+				predicates.ErrNodeLabelPresenceViolated,
+				// this one is probably intentional since it's a workaround for not having
+				// pod hard anti affinity.
+				predicates.ErrPodNotFitsHostPorts,
+				// DaemonSet is expected to respect taints and tolerations
+				predicates.ErrTaintsTolerationsNotMatch:
```

Comment:
This seems wrong. If a DaemonSet doesn't tolerate NoSchedule but the pod is already running, we need to return false, false, true from nodeShouldRunDaemonPod.